src/main/java/core/stack/MateStack.java

```diff
@@ -1,19 +1,53 @@
 package core.stack;
 
+import java.util.EmptyStackException;
+
 public class MateStack<T> {
-    public void push(T value) {
+    private static final int GROWTH_FACTOR = 2;
+    private T[] stack;
+    private int size;
+    private int capacity;
 
+    public MateStack() {
+        final int default_capacity = 10;
+        stack = (T[]) new Object[default_capacity];
+        size = 0;
+        capacity = default_capacity;
     }
 
-    public T peek() {
-        return null;
+    public void push(T value) {
+        if (size >= capacity) {
+            grow();
+        }
+        stack[size] = value;
+        size++;
     }
 
-    public T pop() {
-        return null;
+    public T peek() throws EmptyStackException {
+        if (size == 0) {
+            throw new EmptyStackException();
+        }
+        return stack[size - 1];
+    }
+
+    public T pop() throws EmptyStackException {
+        if (size == 0) {
+            throw new EmptyStackException();
+        }
+        T poppedElement = stack[size - 1];
+        stack[size - 1] = null;
+        size--;
+        return poppedElement;
```

Comment:
The `pop` method correctly throws an `EmptyStackException` when the stack is empty and sets the popped element to `null` to help with garbage collection. Good job on following stack semantics and memory management.